File: inst/Matlab/fdaM/ftpsite/temp/funcint.m
%  ---------------------------------------------------------------

function ss = funcint(fhdl, cvec, basisobj, nderiv1, nderiv2, ...
                     JMAX, EPS)
%  integrates a function of event times and basis values
%  Arguments:
%  FHDL     ... a function handle for the function
%  CVEC     ... a coefficient vector
%  BASISOBJ ... a functional data object for log intensity function w(t)
%  NDERIV1  ... a derivative order 
%  NDERIV2  ... another derivative order

%  Last modified 20 July 2006

if ~strcmp(class(basisobj),'basis')
    error('Fourth argument must be a basis function object.');
end

%  set default arguments

if nargin < 7, EPS  = 1e-7; end
if nargin < 6, JMAX = 9;    end
if nargin < 5, nderiv2 = 0; end
if nargin < 4, nderiv1 = 0; end

rng = getbasisrange(basisobj);

%  set up first iteration

width = rng(2) - rng(1);
JMAXP = JMAX + 1;
h     = ones(JMAXP,1);
h(2) = 0.25;
%  matrix SMAT contains the history of discrete approximations to the integral
%  the first iteration uses just the endpoints
x  = rng';
if nargin == 3
    fval = feval(fhdl, x, cvec, basisobj);
elseif nargin == 4
    fval = feval(fhdl, x, cvec, basisobj, nderiv1);
else
    fval = feval(fhdl, x, cvec, basisobj, nderiv1, nderiv2);
end
scell{1} = width.*fval./2;
tnm = 0.5;

%  now iterate to convergence

for iter = 2:JMAX
    tnm  = tnm*2;
    del  = width/tnm;
    if iter == 2
        x = (rng(1) + rng(2))/2;
    else
        x = (rng(1)+del/2 : del : rng(2))';
    end
    if nargin == 3
        fval = feval(fhdl, x, cvec, basisobj);
    elseif nargin == 4
        fval = feval(fhdl, x, cvec, basisobj, nderiv1);
    else
        fval = feval(fhdl, x, cvec, basisobj, nderiv1, nderiv2);
    end
    scell{iter} = (scell{iter-1} + width.*fval./tnm)./2;
    if iter >= 5
        ind = (iter-4):iter;
        temp = { scell{iter-4}, scell{iter-3}, scell{iter-2}, ...
                 scell{iter-1}, scell{iter} }; 
        [ss, dss] = polintarray(h(ind),temp,0);
        if ~any(abs(dss) > EPS.*max(max(abs(ss))))
            %  successful convergence
            ss = squeeze(ss);
            return;
        end
    end
    scell{iter+1} = scell{iter};
    h(iter+1) = 0.25*h(iter);
end

%  ---------------------------------------------------------------

function [y,dy] = polintarray(xa, ya, x)
%  YA is an array with up to 4 dimensions
%     with 1st dim the same length same as the vector XA
n       = length(xa);
difx    = xa - x;
absxmxa = abs(difx);
tmp = 1:n;
ns = min(tmp(absxmxa == min(absxmxa)));
cs = ya;
ds = ya;
y  = ya{ns};
ns = ns - 1;
for m = 1:(n-1)
    for i = 1:(n-m)
        ho      = difx(i);
        hp      = difx(i+m);
        w       = (cs{i+1} - ds{i})./(ho - hp);
        ds{i} = hp.*w;
        cs{i} = ho.*w;
    end
    if 2*ns < n-m
        dy = cs{ns+1};
    else
        dy = ds{ns};
        ns = ns - 1;
    end
    y = y + dy;
end

File: R/funcint.R
funcint <- function(func,cvec, basisobj, nderiv=0, JMAX=15, EPS=1e-7)
{
  
  #  computes the definite integral of a function defined in terms of 
  #  a functional data object with basis BASISOBJ and coefficient vector CVEC.
  
  #  FUNC     ... the name of a function
  #  CVEC     ..  a vector of coefficients defining the functional data
  #               object required to compute the value of the function.
  #  BASISOBJ ... a functional data basis object
  #  NDERIV   ... a non-negative integer defining a derivative to be used.
  #  JMAX   maximum number of allowable iterations
  #  EPS    convergence criterion for relative stop
  
  #  Return: the value of the function
  
  #  Last modified 15 June 2020 by Jim Ramsay
  
  #  set iter
  
  iter <- 0
  
  # The default case, no multiplicities.
  
  rng    <- basisobj$rangeval
  
  nbasis <- basisobj$nbasis
  
  nrep   <- dim(basisobj$coef)[2]
  
  inprodvec <- matrix(0,nrep,1)
  
  #  set up first iteration
  
  iter  <- 1
  width <- rng[2] - rng[1]
  JMAXP <- JMAX + 1
  h     <- rep(1,JMAXP)
  h[2]  <- 0.25
  s <- matrix(0,JMAXP,nrep)
  sdim <- length(dim(s))
  #  the first iteration uses just the endpoints
  fdobj <- fd(cvec, basisobj)
  fx    <- func(eval.fd(rng, fdobj, nderiv))
  #  multiply by values of weight function if necessary
  s[1,] <- width*apply(fx,2,sum)/2
  tnm  <- 0.5
  
  #  now iterate to convergence
  
  for (iter in 2:JMAX) {
    tnm <- tnm*2
    if (iter == 2) {
      x <- mean(rng)
    } else {
      del <- width/tnm
      x   <- seq(rng[1]+del/2, rng[2]-del/2, del)
    }
    fx <- func(eval.fd(x, fdobj, nderiv))
    chs <- width*apply(fx,2,sum)/tnm
    s[iter,] <- (s[iter-1,] + chs)/2
    if (iter >= 5) {
      ind <- (iter-4):iter
      ya <- s[ind,,]
      ya <- matrix(ya,5,nrep)
      xa <- h[ind]
      absxa <- abs(xa)
      absxamin <- min(absxa)
      ns <- min((1:length(absxa))[absxa == absxamin])
      cs <- ya
      ds <- ya
      y  <- ya[ns,,]
      ns <- ns - 1
      for (m in 1:4) {
        for (i in 1:(5-m)) {
          ho      <- xa[i]
          hp      <- xa[i+m]
          w       <- (cs[i+1,] - ds[i,])/(ho - hp)
          ds[i,] <- hp*w
          cs[i,] <- ho*w
        }
        if (2*ns < 5-m) {
          dy <- cs[ns+1,]
        } else {
          dy <- ds[ns,]
          ns <- ns - 1
        }
        y <- y + dy
      }
      ss     <- y
      errval <- max(abs(dy))
      ssqval <- max(abs(ss))
      if (all(ssqval > 0)) {
        crit <- errval/ssqval
      } else {
        crit <- errval
      }
      if (crit < EPS && iter >= 5) break
    }
    s[iter+1,] <- s[iter,]
    h[iter+1]   <- 0.25*h[iter]
    if (iter == JMAX) warning("Failure to converge.")
  }
  inprodvec <- inprodvec + ss
  
}


